Interrupted connections finish async sequences, addresses discussion #122

jakaplan opened the pull request:
For a given `XPCClient` all active async sequences are terminated with an error if the connection with the server is interrupted. This is the correct behavior because no more responses can possibly be received for this given sequence (although a new requests could be made and might succeed if the server is successfully started once again).

jakaplan committed:
Interrupted connects finish active sequences, addresses discussion #122

For a given `XPCClient` all active async sequences are terminated with an error if the connection with the server is interrupted. This is the correct behavior because no more responses can possibly be received for this given sequence (although a new requests could be made and might succeed if the server is successfully started once again).

jakaplan commented:
@vincent-le-normand I'd appreciate if you could try out this branch and let me know if it addresses your feedback.

jakaplan committed:
Adds test to detect a sequence failing due to server termination

The behavior we want is for the sequence to succeed, but this test is failing. This isn' surprising, but it's unfortunate :(

jakaplan commented:
@vincent-le-normand For more context, I can't commit this as-is because it fails in exactly the way I was worried it would. But before I go a lot deeper and see if I can find a full solution, I wanted to understand if this is addressing your feedback.

vincent-le-normand commented:
Hello,

Yes, this is exactly what I'd need. I test it a little bit more next week, because I got a crash once - but it may be my fault.

Thanks!



> Le 18 nov. 2022 à 13:19, Josh Kaplan ***@***.***> a écrit :
> 
> 
> @vincent-le-normand For more context, I can't commit this as-is because it fails in exactly the way I was worried it would. But before I go a lot deeper and see if I can find a full solution, I wanted to understand if this is addressing your feedback.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.
>

jakaplan committed:
Addresses race conditions during process termination

It's now possible to send a sequential reply which waits on the client to handle it. This allows for a server to terminate its process after knowing the client has received the entire sequence. That said, doing that is not a good approach (this is now documented). However, what this uncovered is this provides a valuable means by which the client can apply back pressure on the server.

This is a breaking change in an async context as there are now new overloaded function names.

Several tests were run in loops 100 times over to ensure the tests reliably pass and are not racing. However, this causes the tests to take a lot longer to complete so they're just running once as reflected in this commit.

jakaplan commented:
@vincent-le-normand I was able to resolve the race conditions. The resulting API changes are a breaking change, so it may be a bit until I make a new official release with this. However, before any of that is relevant I'll await your further testing this upcoming week.

vincent-le-normand commented:
It seems to work, except for one thing. When creating multiple anonymous connection (something we do in unit tests), we are receiving disconnection inappropriately. Here is a small test that is throwing (but shouldn't I think).

`    func testInterruptions() async throws {
		for maxValue in 100 ..< 1000 {
			let route = XPCRoute.named("sequentialRoute").withMessageType(Int.self).withSequentialReplyType(Int.self)
			let server = XPCServer.makeAnonymous()
			let endpoint = server.endpoint
			let client = XPCClient.forEndpoint(endpoint)

			server.registerRoute(route) { maxValue, resultProvider in
				do {
					for value in 0 ..< maxValue {
						try await Task.sleep(nanoseconds: 1_000)
						try await resultProvider.success(value: value)
					}
				}
				catch {
					try? await resultProvider.failure(error: error)
				}
			}

			server.start()

			let asyncStream = client.sendMessage(maxValue, to: route)
			_ = try await asyncStream.first { _ in false }
		}
	}`

jakaplan commented:
@vincent-le-normand I looked at this briefly and can replicate this. You may be correct that there's a subtle race condition here which I'll need to investigate more. However, the code you've written maximizes the chance of this due to never finishing a sequence. If I add: `try await resultProvider.finished()` after:
```
for value in 0 ..< maxValue {
    try await Task.sleep(nanoseconds: 1_000)
    try await resultProvider.success(value: value)
}
```
then the test always passes for me. This is not to say a bug does not exist within the SecureXPC package.

I don't expect to have more time to investigate this week as I'm on holiday in New Caledonia, but should have time next week to look into this further.

vincent-le-normand commented:
Thanks for the reply.

You're right, it's much better when adding .finished after the for loop.

According to what I understand of the code, the sequence should be automatically finished when the resultProvider is dented. So the race condition only happens in this case - at least for this particular example.

No rush here. It's really appreciated that you take a look to this feature.

I'll probably have a couple of other feature i would like to add - ill make PR for them.

One is related to testing this interrupted server case. It would be nice to be able to programatically interrupt a XPCServer.

Another one is not directly related, but it would be useful to bu able to be notified globally when interruption handler is called. For instance, you can adapt the user interface if a daemon crashes or is unloaded by new ventura settings. 

Enjoy your vacations.

Vincent

> Le 23 nov. 2022 à 13:19, Josh Kaplan ***@***.***> a écrit :
> 
> 
> @vincent-le-normand I looked at this briefly and can replicate this. You may be correct that there's a subtle race condition here which I'll need to investigate more. However, the code you've written maximizes the chance of this due to never finishing a sequence. If I add: try await resultProvider.finished() after:
> 
> for value in 0 ..< maxValue {
>     try await Task.sleep(nanoseconds: 1_000)
>     try await resultProvider.success(value: value)
> }
> 
> then the test always passes for me. This is not to say a bug does not exist within the SecureXPC package.
> 
> I don't expect to have more time to investigate this week as I'm on holiday in New Caledonia, but should have time next week to look into this further.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.
>

jakaplan commented:
> According to what I understand of the code, the sequence should be automatically finished when the resultProvider is dented. So the race condition only happens in this case - at least for this particular example.

These are not equivalent operations. When the resulted provider is deinitialized a message is sent from the server to the client to finish the sequence is finished, but no ordering is guaranteed server side as nothing is `await`ed. Whereas calling `try await resultProvider.finished()` means that the server code is waiting for client confirmation before proceeding.

I'm back from my holiday now and should have sometime in the next few days to debug this more.

jakaplan committed:
Ends transaction during provider deinit

This addresses a discrepancy in behavior between when an API user explicitly finished a sequence vs when it's done as part of deinitialization.

jakaplan commented:
@vincent-le-normand I believe I've found and resolved the issue. With the latest changes, I can run your unit test without issue. Let me know if anything continues to not behave as expected.

(What I described above was correct in describing the problem, but it turns out there was actually a much more straightforward solution than I had anticipated - there was actually a bug in terms of how deinit's finishing of a sequence behaved.)

vincent-le-normand commented:
Everything seems to work as I would expect so far. Thanks a lot!
